File: algorithms/python/minimumNumbersofFunctionCallstoMakeTargetArray/minimumNumbersofFunctionCallstoMakeTargetArray.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/minimum-numbers-of-function-calls-to-make-target-array/
# Author: Miao Zhang
# Date:   2021-05-15

class Solution:
    def minOperations(self, nums: List[int]) -> int:
        return sum(bin(x).count('1') for x in nums) + len(bin(max(nums))) - 3

File: algorithms/cpp/minimumNumbersofFunctionCallstoMakeTargetArray/minimumNumbersofFunctionCallstoMakeTargetArray.cpp
// Source: https://leetcode.com/problems/minimum-numbers-of-function-calls-to-make-target-array/
// Author: Miao Zhang
// Date:   2021-05-15

class Solution {
public:
    int minOperations(vector<int>& nums) {
        int res = 0;
        int high = 0;
        for (int x: nums) {
            high = max(high, 31 - __builtin_clz(x | 1));
            res += std::bitset<32>(x).count();
        }
        return res + high;
    }
};
